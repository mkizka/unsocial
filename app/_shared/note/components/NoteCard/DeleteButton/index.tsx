"use client";
import { XMarkIcon } from "@heroicons/react/24/outline";
import { useTransition } from "react";

import { Spinner } from "@/_shared/ui/Spinner";

type Props = {
  onClick: () => Promise<void>;
};

export function DeleteButton({ onClick }: Props) {
  const [isPending, startTransition] = useTransition();
  return (
    <button
      data-testid="delete-button"
      onClick={() => startTransition(onClick)}
      className="size-5 text-accent transition-colors"
    >
      {isPending ? <Spinner /> : <XMarkIcon />}
    </button>
  );
}
