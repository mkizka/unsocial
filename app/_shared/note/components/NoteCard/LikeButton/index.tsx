"use client";
import { HeartIcon as UnLikedIcon } from "@heroicons/react/24/outline";
import { HeartIcon as LikedIcon } from "@heroicons/react/24/solid";
import { useState } from "react";

import { Button } from "@/_shared/ui/Button";

type Props = {
  isLiked: boolean;
  likesCount: number;
  onClick: () => Promise<void>;
};

export function LikeButton({
  isLiked: initialIsLiked,
  likesCount: initialLikesCount,
  onClick,
}: Props) {
  // 表示上は即時反映させる
  const [isLiked, setIsLiked] = useState(initialIsLiked);
  const [likesCound, setLikesCount] = useState(initialLikesCount);
  return (
    <Button
      variant="ghost"
      size="icon"
      data-testid="like-button"
      type="button"
      onClick={async () => {
        setIsLiked((prev) => !prev);
        setLikesCount((prev) => (isLiked ? prev - 1 : prev + 1));
        await onClick();
      }}
    >
      {isLiked ? (
        <LikedIcon className="size-5 text-primary transition-colors" />
      ) : (
        <UnLikedIcon className="size-5 text-primary transition-colors" />
      )}
      <div className="absolute left-8">{likesCound}</div>
    </Button>
  );
}
