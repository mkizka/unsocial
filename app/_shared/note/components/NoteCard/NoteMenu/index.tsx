"use client";
import { Bars3Icon, HeartIcon, XMarkIcon } from "@heroicons/react/24/outline";
import { useState } from "react";

import { Button } from "@/_shared/ui/Button";
import { Card } from "@/_shared/ui/Card";
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "@/_shared/ui/Dialog";

type Props = {
  noteId: string;
};

function LikesDialog() {
  return (
    <Dialog>
      <DialogTrigger asChild>
        <button className="flex items-center gap-2 hover:opacity-70">
          <HeartIcon className="size-5" />
          <div>いいねしたユーザー</div>
        </button>
      </DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>いいねしたユーザー</DialogTitle>
          <DialogDescription></DialogDescription>
        </DialogHeader>
      </DialogContent>
    </Dialog>
  );
}

export function NoteMenu({ noteId }: Props) {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <div className="relative ml-auto size-9">
      <Button
        variant="ghost"
        size="icon"
        data-testid="like-button"
        type="button"
        onClick={() => setIsOpen((prev) => !prev)}
      >
        {isOpen ? (
          <XMarkIcon className="size-5 transition-colors" />
        ) : (
          <Bars3Icon className="size-5 transition-colors" />
        )}
      </Button>
      {isOpen && (
        <>
          <Card
            className="absolute right-0 z-10 w-56 space-y-4 drop-shadow-xl"
            data-testid="note-menu__dropdown"
          >
            <LikesDialog />
          </Card>
          <div
            className="fixed inset-0 z-0"
            onClick={() => setIsOpen(false)}
            data-testid="note-menu__backdrop"
          />
        </>
      )}
    </div>
  );
}
