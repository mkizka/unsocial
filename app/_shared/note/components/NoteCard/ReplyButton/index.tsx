"use client";

import { ChatBubbleLeftIcon } from "@heroicons/react/24/outline";
import Link from "next/link";

type Props = {
  url: string;
};

export function ReplyButton({ url }: Props) {
  return (
    <Link href={url + "?reply"} data-testid="note-card__reply">
      <ChatBubbleLeftIcon className="size-5 text-dark transition-colors hover:text-gray" />
    </Link>
  );
}
