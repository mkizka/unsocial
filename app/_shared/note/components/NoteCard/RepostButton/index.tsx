"use client";
import { ArrowPathIcon } from "@heroicons/react/24/outline";
import { useState } from "react";

import { Button } from "@/_shared/ui/Button";
import { cn } from "@/_shared/utils/cn";

type Props = {
  isReposted: boolean;
  onClick: () => Promise<void>;
};

export function RepostButton({ isReposted: initialState, onClick }: Props) {
  // 表示上は即時反映させる
  const [isReposted, setIsReposted] = useState(initialState);
  return (
    <Button
      variant="ghost"
      size="icon"
      data-testid="repost-button"
      type="button"
      onClick={async () => {
        setIsReposted((prev) => !prev);
        await onClick();
      }}
    >
      <ArrowPathIcon
        className={cn("size-5 transition-colors", {
          "text-primary": isReposted,
        })}
      />
    </Button>
  );
}
