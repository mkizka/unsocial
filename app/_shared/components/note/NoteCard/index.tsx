import { action as deleteAction } from "./DeleteButton/action";
import { action as likeAction } from "./LikeButton/action";
import { NoteCard, type NoteCardProps } from "./NoteCard";

type Props = Omit<NoteCardProps, "onLike" | "onDelete">;

function NoteCardContainer(props: Props) {
  console.log(props);
  return (
    <NoteCard
      {...props}
      onLike={() =>
        likeAction({
          noteId: props.note.id,
          content: "👍",
        })
      }
      onDelete={() => deleteAction(props.note.id)}
    />
  );
}

export { NoteCardContainer as NoteCard };
