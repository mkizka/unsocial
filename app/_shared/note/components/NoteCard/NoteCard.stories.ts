import type { Meta, StoryObj } from "@storybook/react";

import type { noteCardFindService } from "@/_shared/note/services/noteCardFindService";

import { NoteCard } from "./NoteCard";

const note: noteCardFindService.NoteCard = {
  id: "1",
  publishedAt: new Date("2024-01-01T00:00:00.000Z"),
  content: "投稿内容",
  attachmentUrls: [],
  url: "/notes/1",
  isLiked: false,
  isMine: false,
  // @ts-expect-error
  user: {
    url: "/user",
    name: "ユーザー名",
    displayUsername: "@preferredUsername@example.com",
    iconHash: "default",
  },
};

const meta: Meta<typeof NoteCard> = {
  component: NoteCard,
  args: {
    note,
  },
};

export default meta;

type Story = StoryObj<typeof NoteCard>;

export const Default: Story = {};

export const Repost: Story = {
  args: {
    note: {
      ...note,
      quotedBy: note.user,
    },
  },
};
