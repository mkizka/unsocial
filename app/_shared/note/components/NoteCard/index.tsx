"use client";
import { action as deleteAction } from "./DeleteButton/action";
import { action as likeAction } from "./LikeButton/action";
import { NoteCard, type NoteCardProps } from "./NoteCard";
import { action as repostAction } from "./RepostButton/action";

type Props = Omit<NoteCardProps, "onRepost" | "onLike" | "onDelete">;

export function NoteCardContainer(props: Props) {
  return (
    <NoteCard
      {...props}
      onRepost={() => repostAction({ noteId: props.note.id })}
      onLike={() => likeAction({ noteId: props.note.id, content: "👍" })}
      onDelete={() => deleteAction(props.note.id)}
    />
  );
}
