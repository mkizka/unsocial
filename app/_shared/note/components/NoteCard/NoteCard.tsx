import Link from "next/link";
import sanitizeHtml from "sanitize-html";

import type { noteCardFindService } from "@/_shared/note/services/noteCardFindService";
import { Card } from "@/_shared/ui/Card";
import { UserIcon } from "@/_shared/user/components/UserIcon";
import { cn } from "@/_shared/utils/cn";

import { AttachmentImages } from "./AttachmentImages";
import { DeleteButton } from "./DeleteButton";
import { LikeButton } from "./LikeButton";
import { PublishedAt } from "./PublishedAt";
import { ReplyButton } from "./ReplyButton";

export type NoteCardProps = {
  note: noteCardFindService.NoteCard;
  onLike: () => Promise<void>;
  onDelete: () => Promise<void>;
  showDetail?: boolean;
  withReplyLink?: boolean;
};

export function NoteCard({
  note,
  onLike,
  onDelete,
  showDetail,
  withReplyLink,
}: NoteCardProps) {
  return (
    <Card
      as="article"
      data-testid="note-card"
      className={cn("text-prmary relative space-y-2", {
        "space-y-4": showDetail,
      })}
    >
      <div className="flex pl-[48px]">
        <Link className="absolute left-3" href={note.user.url}>
          <UserIcon user={note.user} className="rounded-full" size={36} />
        </Link>
        <div className={cn("w-full space-y-2", { "space-y-4": showDetail })}>
          <div className="flex">
            <Link href={note.user.url} className="truncate hover:underline">
              {note.user.name && (
                <span className="mr-1 font-bold">{note.user.name}</span>
              )}
              <span className="text-gray">{note.user.displayUsername}</span>
            </Link>
            <PublishedAt href={note.url} publishedAt={note.publishedAt} />
          </div>
          <div
            className={cn("break-words", { "text-xl": showDetail })}
            dangerouslySetInnerHTML={{ __html: sanitizeHtml(note.content) }}
          ></div>
          <AttachmentImages urls={note.attachmentUrls} />
        </div>
      </div>
      <footer className="mt-3 space-y-2 pl-[48px]">
        <div className="flex gap-10">
          <ReplyButton url={note.url} />
          <LikeButton isLiked={note.isLiked} onClick={onLike} />
          {note.isMine && <DeleteButton onClick={onDelete} />}
        </div>
        {withReplyLink && (
          <Link
            className="block text-gray hover:underline"
            href={`/notes/${note.id}`}
          >
            返信を表示
          </Link>
        )}
      </footer>
    </Card>
  );
}
