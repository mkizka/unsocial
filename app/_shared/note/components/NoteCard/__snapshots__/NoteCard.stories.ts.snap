// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_shared/note/components/NoteCard Default smoke-test 1`] = `
<article class="rounded-lg border bg-card p-4 text-card-foreground shadow-sm relative space-y-2"
         data-testid="note-card"
>
  <div class="flex pl-[48px]">
    <a class="absolute left-3"
       href="/user"
    >
      <img class="hover:opacity-80 rounded-full"
           width="36"
           height="36"
           src="/icons/default.webp?size=36"
           alt="@undefinedのアイコン"
      >
    </a>
    <div class="w-full space-y-2">
      <div class="flex">
        <a class="truncate hover:underline"
           href="/user"
        >
          <span class="mr-1 font-bold">
            ユーザー名
          </span>
          <span>
            @preferredUsername@example.com
          </span>
        </a>
        <a data-testid="note-card__link"
           class="ml-auto whitespace-nowrap hover:underline"
           href="/notes/1"
        >
          2024/1/1
        </a>
      </div>
      <div class="break-words">
        投稿内容
      </div>
    </div>
  </div>
  <footer class="space-y-2 pl-[48px]">
    <div class="flex gap-10">
      <a data-testid="note-card__reply"
         href="/notes/1?reply"
      >
        <button class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-secondary size-8"
                type="button"
        >
          <span class>
            <svg xmlns="http://www.w3.org/2000/svg"
                 fill="none"
                 viewbox="0 0 24 24"
                 stroke-width="1.5"
                 stroke="currentColor"
                 aria-hidden="true"
                 data-slot="icon"
                 class="size-5 transition-colors"
            >
              <path stroke-linecap="round"
                    stroke-linejoin="round"
                    d="M2.25 12.76c0 1.6 1.123 2.994 2.707 3.227 1.087.16 2.185.283 3.293.369V21l4.076-4.076a1.526 1.526 0 0 1 1.037-.443 48.282 48.282 0 0 0 5.68-.494c1.584-.233 2.707-1.626 2.707-3.228V6.741c0-1.602-1.123-2.995-2.707-3.228A48.394 48.394 0 0 0 12 3c-2.392 0-4.744.175-7.043.513C3.373 3.746 2.25 5.14 2.25 6.741v6.018Z"
              >
              </path>
            </svg>
          </span>
        </button>
      </a>
      <button class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-secondary size-8"
              data-testid="repost-button"
              type="button"
      >
        <span class>
          <svg xmlns="http://www.w3.org/2000/svg"
               fill="none"
               viewbox="0 0 24 24"
               stroke-width="1.5"
               stroke="currentColor"
               aria-hidden="true"
               data-slot="icon"
               class="size-5 transition-colors"
          >
            <path stroke-linecap="round"
                  stroke-linejoin="round"
                  d="M16.023 9.348h4.992v-.001M2.985 19.644v-4.992m0 0h4.992m-4.993 0 3.181 3.183a8.25 8.25 0 0 0 13.803-3.7M4.031 9.865a8.25 8.25 0 0 1 13.803-3.7l3.181 3.182m0-4.991v4.99"
            >
            </path>
          </svg>
        </span>
      </button>
      <button class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-secondary size-8"
              data-testid="like-button"
              type="button"
      >
        <span class>
          <svg xmlns="http://www.w3.org/2000/svg"
               fill="none"
               viewbox="0 0 24 24"
               stroke-width="1.5"
               stroke="currentColor"
               aria-hidden="true"
               data-slot="icon"
               class="size-5 text-primary transition-colors"
          >
            <path stroke-linecap="round"
                  stroke-linejoin="round"
                  d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12Z"
            >
            </path>
          </svg>
        </span>
      </button>
    </div>
  </footer>
</article>
`;

exports[`_shared/note/components/NoteCard Repost smoke-test 1`] = `
<article class="rounded-lg border bg-card p-4 text-card-foreground shadow-sm relative space-y-2"
         data-testid="note-card"
>
  <div class="flex items-center pl-[48px]">
    <svg xmlns="http://www.w3.org/2000/svg"
         fill="none"
         viewbox="0 0 24 24"
         stroke-width="1.5"
         stroke="currentColor"
         aria-hidden="true"
         data-slot="icon"
         class="absolute left-8 h-5 text-primary"
    >
      <path stroke-linecap="round"
            stroke-linejoin="round"
            d="M16.023 9.348h4.992v-.001M2.985 19.644v-4.992m0 0h4.992m-4.993 0 3.181 3.183a8.25 8.25 0 0 0 13.803-3.7M4.031 9.865a8.25 8.25 0 0 1 13.803-3.7l3.181 3.182m0-4.991v4.99"
      >
      </path>
    </svg>
    <a class="hover:underline"
       href="/user"
    >
      ユーザー名さん
    </a>
    がリポストしました
  </div>
  <div class="flex pl-[48px]">
    <a class="absolute left-3"
       href="/user"
    >
      <img class="hover:opacity-80 rounded-full"
           width="36"
           height="36"
           src="/icons/default.webp?size=36"
           alt="@undefinedのアイコン"
      >
    </a>
    <div class="w-full space-y-2">
      <div class="flex">
        <a class="truncate hover:underline"
           href="/user"
        >
          <span class="mr-1 font-bold">
            ユーザー名
          </span>
          <span>
            @preferredUsername@example.com
          </span>
        </a>
        <a data-testid="note-card__link"
           class="ml-auto whitespace-nowrap hover:underline"
           href="/notes/1"
        >
          2024/1/1
        </a>
      </div>
      <div class="break-words">
        投稿内容
      </div>
    </div>
  </div>
  <footer class="space-y-2 pl-[48px]">
    <div class="flex gap-10">
      <a data-testid="note-card__reply"
         href="/notes/1?reply"
      >
        <button class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-secondary size-8"
                type="button"
        >
          <span class>
            <svg xmlns="http://www.w3.org/2000/svg"
                 fill="none"
                 viewbox="0 0 24 24"
                 stroke-width="1.5"
                 stroke="currentColor"
                 aria-hidden="true"
                 data-slot="icon"
                 class="size-5 transition-colors"
            >
              <path stroke-linecap="round"
                    stroke-linejoin="round"
                    d="M2.25 12.76c0 1.6 1.123 2.994 2.707 3.227 1.087.16 2.185.283 3.293.369V21l4.076-4.076a1.526 1.526 0 0 1 1.037-.443 48.282 48.282 0 0 0 5.68-.494c1.584-.233 2.707-1.626 2.707-3.228V6.741c0-1.602-1.123-2.995-2.707-3.228A48.394 48.394 0 0 0 12 3c-2.392 0-4.744.175-7.043.513C3.373 3.746 2.25 5.14 2.25 6.741v6.018Z"
              >
              </path>
            </svg>
          </span>
        </button>
      </a>
      <button class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-secondary size-8"
              data-testid="repost-button"
              type="button"
      >
        <span class>
          <svg xmlns="http://www.w3.org/2000/svg"
               fill="none"
               viewbox="0 0 24 24"
               stroke-width="1.5"
               stroke="currentColor"
               aria-hidden="true"
               data-slot="icon"
               class="size-5 transition-colors"
          >
            <path stroke-linecap="round"
                  stroke-linejoin="round"
                  d="M16.023 9.348h4.992v-.001M2.985 19.644v-4.992m0 0h4.992m-4.993 0 3.181 3.183a8.25 8.25 0 0 0 13.803-3.7M4.031 9.865a8.25 8.25 0 0 1 13.803-3.7l3.181 3.182m0-4.991v4.99"
            >
            </path>
          </svg>
        </span>
      </button>
      <button class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-secondary size-8"
              data-testid="like-button"
              type="button"
      >
        <span class>
          <svg xmlns="http://www.w3.org/2000/svg"
               fill="none"
               viewbox="0 0 24 24"
               stroke-width="1.5"
               stroke="currentColor"
               aria-hidden="true"
               data-slot="icon"
               class="size-5 text-primary transition-colors"
          >
            <path stroke-linecap="round"
                  stroke-linejoin="round"
                  d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12Z"
            >
            </path>
          </svg>
        </span>
      </button>
    </div>
  </footer>
</article>
`;
